Feature interactive dot graph

Kanakanajm opened the pull request:
### New Features (for interactive mode only)
The dot graph image for each proof step is now interactable. They are added onto the DOM as web component. Here are the new features of the graph:
- Zoom in and out with different display settings (protocol nodes got simplified)
- Pan
- Highlight graph nodes when clicking on the items of the abbreviation table
- Highlight graph path when clicking on nodes 
- Pop into new tab (the content and graph options are synchronized with the host window i.e. when user navigate through the proof steps, the content on the popup window changes accordingly). User can put the pop window on an external display and navigate on host window.
- Hint (not feature): user can still Ctrl-F to search and copy the text on the graph nodes.

### What has been changed?
- A new `frontend` folder that contains the source code of the web component
- Under `src/Web` 
  - A new endpoint to serve dot graph string (previously as `*.dot` file) on requests
  - Change on templates instead of returning image component now return `dot-graph-viz` web component
  - The interactive graph is available for both normal and diff mode (with `--diff` argument)

### Contributing Guideline Comments
We compared the result of the regression tests from the `develop` branch and our branch which are identical.
We carried out the steps from the first bullet point of "Making new releases" section. The result of `iii` are also identical for both branches.

### Licenses of frontend dependencies:
- viz.js: MIT
- d3.js: ISC
Both are compatible with [GNU GPL](https://www.gnu.org/licenses/gpl-faq.html#WhatDoesCompatMean).

Kanakanajm committed:
Scaffold ts project

Kanakanajm committed:
Add a basic example of render dot graph w/ d3-graphviz

Kanakanajm committed:
Add viz-js example and it zooms correctly

Kanakanajm committed:
Remove unused transform fn

Kanakanajm committed:
Add cursor zoom but it has a slight shift after cursor change position

Kanakanajm committed:
Add zoom on centroid only option

Kanakanajm committed:
One line fix, new viz-js can also zoom

Kanakanajm committed:
Add zoom out ellipse highlight

Kanakanajm committed:
Correct example dot typo

Kanakanajm committed:
Add zoom out reshape

Kanakanajm committed:
First version of reroute edge

Kanakanajm committed:
Clean up codes

Kanakanajm committed:
Refactor dot graph visualization into web component and define DiGraph for storing graph connection info

Kanakanajm committed:
Add highlight on click

Kanakanajm committed:
Fix previous highlight not cleared bug

Kanakanajm committed:
Add template for abbrevation highlight

Kanakanajm committed:
Add abbreviation highlight

Kanakanajm committed:
Remove unnecessary logs

Kanakanajm committed:
Fix zoom out highlighted class set bug

Divya19gupta committed:
added comment

Kanakanajm committed:
Remove unused function

Divya19gupta committed:
added ui fixes to center the graph, added max and min zoom values and also merged the copy/find feature

Divya19gupta committed:
remove selection text when clicked outside the graph

Commit:
Add frontend module

Commit:
Fix css style override and random chunk name

Commit:
Deploy new frontend

Kanakanajm committed:
Fix click event not bound bug and edge end not projected bug

Commit:
Temporarily removed submodule

Commit:
Add submodule back

Kanakanajm committed:
Remove unused const

Kanakanajm committed:
Add new node shape

Kanakanajm committed:
Fix edges undefined bug

Kanakanajm committed:
Try fix multiple graph interference

Kanakanajm committed:
Remove fixed dev graph

Commit:
Fix unclosed tag leak problem

Commit:
Build with latest frontend

Kanakanajm committed:
Separate prod and dev dot string load mechanism

Kanakanajm committed:
Add popup and url sync

Kanakanajm committed:
Try use broadcast channel instead

Kanakanajm committed:
Add url sync for channel

Commit:
Make sync initiated from popup window

Kanakanajm committed:
Fix abbreviation box position and display graph as block

Kanakanajm committed:
Fix projected arrow head wrong direction bug

Commit:
Fix pop up url

Commit:
Hide popup button for raw and refined sources

Commit:
Enable popup function and fix abbreviation

Commit:
Fix options not applied to the dot graph src bug

Kanakanajm committed:
Increase record font size on zoom out

Divya19gupta committed:
Added fix for white background and text issue for popin/out

Kanakanajm committed:
Move render logic outside of component connected callback to enable reusablility

Kanakanajm committed:
Sync dotsrc instead of window href

Commit:
Fix option not sync on popup window bug

Kanakanajm committed:
Keep shape during zoom out and abbreviate and change font size of records

Kanakanajm committed:
Scale up abbreviation size and set threshold for font resize

Commit:
Fix action text not shorten properly without square bracket

Commit:
Bump frontend

Commit:
fix for previous nodes are visible even when new option is selected

Commit:
fixed zoom level

Commit:
Add interactive dot for diff

Kanakanajm committed:
Remove frontend submodule

Kanakanajm committed:
Add 'frontend/' from commit 'dbf1e186c08ba52348542189cacb416616de70b5'

git-subtree-dir: frontend
git-subtree-mainline: 12fc91ea5e46a75aee93c49e211b5c73059ea27f
git-subtree-split: dbf1e186c08ba52348542189cacb416616de70b5

Commit:
Pack and minor change

Commit:
Extract functions related to interactive dot graph and set no popup on source

Commit:
Merge remote-tracking branch 'origin/develop' into feature-interactive-dot-graph

Commit:
Remove duplicate

Commit:
Fix merge bug

Commit:
added comments and fixed frontend issues

Commit:
Fix Options synchronization issue

rkunnema committed:
Merge branch 'develop' into feature-interactive-dot-graph

Porrum commented:
Hi all,

The features you implemented work like a charm! I do have a few comments regarding User Experience, though.
1. The `Pop-out` button is a bit hidden and kind of detached from the other graph options; I think it would fit better as a toggle in the `Options` menu.
1. If you try to scroll down in the right panel of the UI, you now often zoom into the graph instead. Since the border around the graph is gone, it is not visually clear where the cursor has to be if you want to scroll.
1. When the graph view is popped out, there is a lot of whitespace in the main window.

Commit:
Add border and compact space when popout

Kanakanajm commented:
> Hi all,
> 
> The features you implemented work like a charm! I do have a few comments regarding User Experience, though.
> 
> 1. The `Pop-out` button is a bit hidden and kind of detached from the other graph options; I think it would fit better as a toggle in the `Options` menu.
> 2. If you try to scroll down in the right panel of the UI, you now often zoom into the graph instead. Since the border around the graph is gone, it is not visually clear where the cursor has to be if you want to scroll.
> 3. When the graph view is popped out, there is a lot of whitespace in the main window.

Hi Eric, thanks for the comments.
I gave the graph a black border and now the whitespace are gone after popping out.
About the `Pop-out` button: if it is added in graph options, when the toggle is on, for those pages with case studies (more than one graph on the same page) should i render all the graphs in the pop out window or only the constraint system graph?

Kanakanajm commented:
Sorry, the force-pushed is because I accidentally pushed some new changes into this PR and reset the commit. Please ignore it.

Porrum commented:
Looks great, the border is really helpful!

Ideally, the user would be able to configure which graphs on given page should be popped out to a separate window and which are not. I could imagine a dropdown menu within the Options menu where you can either pop out all graphs, or toggle a pop-out or pop-in for every graph.

Porrum commented:
Another thing I just noticed: The graphs both for Raw and Refined Sources cannot be popped out (in that they do not have an associated `Pop-out` button), but when I pop a constraint graph from any lemma out and switch to Sources, the source graphs are also popped out. They can be popped in by clicking any of the `Pop-in` buttons that are visible, after which no `Pop-out` button appears.

Commit:
Add static popup, fix dynamic popup fail to detect closed bug and restructure frontend project folder

Porrum commented:
Looks good to me now (modulo rebase), thanks a lot!

Kanakanajm committed:
Merge branch 'develop' into feature-interactive-dot-graph

cascremers commented:
Okay, it seems to work pretty well on my machine! Some remaining questions:
 - If I pop out a window that is more portrait than landscape, the abbreviation legend is not in the right bottom. (See attached screenshot). Can we fix this?
 - In the previous trial implementations, we had three zoom levels. Is there any reason why we lost one?

<img width="540" height="946" alt="Screenshot 2025-10-24 at 10 12 59" src="https://github.com/user-attachments/assets/3cacdfb1-b95f-4702-a582-6b90698c2680" />

cascremers commented:
Another question: currently, the criterion when to switch zoom level seems to be only based on whether the user zooms out the graph to be smaller than the display window.

I would prefer that we determine the zoom level based on the display font size: when text is smaller than a specific threshold, we switch to the other zoom level. This should then ensure that the rule names are (nearly) always visible when opening a graph, no matter how large.

Kanakanajm committed:
Fix static graph template graph source

Kanakanajm committed:
Fix abbreviation table position to the bottom right corner of the svg container (WIP,  re-render needed when resize handle triggers)

Kanakanajm committed:
Use ResizeObserver to detect svg container size change and trigger re-position for abbreviation table

Commit:
Build frontend

Kanakanajm commented:
> Okay, it seems to work pretty well on my machine! Some remaining questions:
> 
> * If I pop out a window that is more portrait than landscape, the abbreviation legend is not in the right bottom. (See attached screenshot). Can we fix this?

This point is fixed now. I fixed the table at the bottom right corner of the svg container with a 10 unit margin.

Kanakanajm committed:
Zoom level determined by pixel coverage (frontend preview)

Commit:
Build frontend

Kanakanajm commented:
> Another question: currently, the criterion when to switch zoom level seems to be only based on whether the user zooms out the graph to be smaller than the display window.
> 
> I would prefer that we determine the zoom level based on the display font size: when text is smaller than a specific threshold, we switch to the other zoom level. This should then ensure that the rule names are (nearly) always visible when opening a graph, no matter how large.

Now abstraction level is determined by how many pixels is covered by 1 unit in SVG coordinate system.

Kanakanajm committed:
Solve incorrect legend position on firefox

Firefox gives null for calling getCTM() on svg container.\nThus now the scaling and tranlation is calculated manually to map screen coordinate into user coordinate without using matrix

Kanakanajm committed:
Legend now has a default font size of 12px on screen no matter how big the graph is + build frontend

Kanakanajm committed:
Abstraction level switch now depends on if text font size on screen goes over 12px + build front

Kanakanajm committed:
Add graph popout hint to manual

Commit:
reformatted, updated the note on window pop-outs in the manual

Porrum committed:
minor rewrite of graph pop-out paragraph in manual

felixlinker commented on frontend/src/wcs/popupgraph.ts:
Thy does this empty file exist?

felixlinker commented on .gitmodules:
Why does this file exist if it's empty? Doesn't seem like you're using git modules. Would be great if this could be removed.

felixlinker commented on data/js/tamarin-prover-ui.js:
Apart from like 5 lines, all changes are in whitespaces. Did you use some autoformatter? Are they required by typescript? I'm not the biggest fan of unneccessary diffs.

felixlinker commented on frontend/index.html:
Is the purpose of this file (and the dynamic ones, etc.) for testing or demonstration?

Kanakanajm replied:
Yes

felixlinker commented on data/js/intdot-dynamicgraph.es.js:
I don't particularly like that these files exist. I think it would be better to gitignore them and only depend on make. Anything speaking against this?

Kanakanajm replied:
The frontend functionalities are written as web components and are later built and bundled by Vite in library mode. They act as a library for tamarin since all of tamarin's interactive mode pages are served by tamarin itself. The frontend project only supplies components to be consumed. It is of the same reason why `data/js/jquery.js` exists.

felixlinker reviewed (changes requested):
Thanks for this awesome feature! I did not look at all Javascript code in detail, but you seem to know what you're doing. I played around a bit with the new feature and it seems to work quite well.

I only have some small things that I'd ideally like to see changed.
- You seem to have some example or testing code in the frontend folder, e.g., the HTML files. I think it would be much cleaner and help future contributors to have a library folder where all the code lives that will be used by Tamarin, and another folder, e.g., called "example" where you use the code from the library.
- I would .gitignore the compiled files in `data/js` and rely on Make only. Committing these compiled files can easily lead to git issues.
- There are some empty files, and I wonder why they exist. Would be good to delete them if possible.

These are the more important asks. I may have brought up some simpler things in the comments that I forgot in my summary now :)

Commit:
removed unnecessary files

Kanakanajm committed:
Revert auto format on ui file

Kanakanajm committed:
Remove cjs format builds and static asset

Kanakanajm committed:
Relocate example web pages

Kanakanajm committed:
Merge branch 'develop' into feature-interactive-dot-graph

felixlinker committed:
Ignore changes to compiled .js files

felixlinker committed:
Remove unneccessary dependency

felixlinker reviewed (changes requested):
Thanks! We're almost there :) I still think it would be best to not track compiled JavaScript files. In my opinion, it would be best to fully rely on make. I created a PR to your PR 😁 https://github.com/Kanakanajm/tamarin-prover-cc-dev/pull/24 Would be get to get another pair of eyes on this, but if you agree that this is better handling, we're good to go form my side!

Kanakanajm committed:
Merge pull request #24 from felixlinker/feature-interactive-dot-graph

Ignore changes to compiled .js files

felixlinker reviewed (approved):
Thanks for your work!

cascremers commented:
Thanks all, this looks very good now. Merging into develop.
